testing_system/code_extractor/code_extractor.py

```diff
@@ -0,0 +1,295 @@
+                        pass
+            else:
+                self.continue_traverse(self.traverse_collect_model_info, node, cell_no)
+
+        def traverse_fill_missing_fit_loc(self, node: ast.AST, cell_no: int):
+            if not isinstance(node, ast.AST):
+                return
+
+            missing_fit_loc_models = [mi for mi in self.model_info if mi.fit_location is None]
+
+            if len(missing_fit_loc_models) == 0:
+                return
+
+            if isinstance(node, ast.Call) and get_func_name(node) == 'fit':
+                try:
+                    mi = next((mi for mi in missing_fit_loc_models if mi.variable_name == node.func.value.id), None)
+                    if mi is not None:
+                        mi.fit_location = NtbTokenLocationInfo(cell_no, node.lineno)
+                        mi.X_train_var_name = node.args[0].id
+                        mi.y_train_var_name = node.args[1].id
+                except AttributeError:
+                    pass
+            else:
+                self.continue_traverse(self.traverse_fill_missing_fit_loc, node, cell_no)
+
+    def get_model_info(self):
+        return self.model_info
+
+    def get_model_names(self):
+        return [mi.model_name for mi in self.model_info]
+
+    def get_preprocessing_code(self) -> str:
+        """
+        By default returns code above constructor of first encountered model
+        """
+        first_model = self.model_info[0]
+        first_model_cell_no, first_model_line_no = first_model.constructor_location.cell_no, \
+                                                   first_model.constructor_location.line_no - 1
+
+        return NtbCodeExtractorHelper.generate_code_up_to(first_model_cell_no,
+                                                          first_model_line_no,
+                                                          path_to_ntb_file=self.path_to_ntb_file,
+                                                          including_last_line=False)
+
+    def get_preprocessing_script(self) -> str:
+        first_model = self.model_info[0]
+
+        res = ''
+        res += self.get_preprocessing_code()
+        res += self.__generate_preproc_return_stmt(first_model.X_train_var_name, first_model.y_train_var_name)
+
+        return res
+
+    def generate_training_script(self, model_name=None, how_to_pickle='pickle') -> str:
+        if model_name is not None:
+            encountered_model_names = self.get_model_names()
+            if model_name not in encountered_model_names:
+                raise RuntimeError(f'Model {model_name} was not found in {self.path_to_ntb_file} notebook.')
+            if model_name != encountered_model_names[0]:
+                raise RuntimeError(f'Not Implemented. Required model {model_name} is not the first to occur'
+                                   f'in {self.path_to_ntb_file} notebook.')
+
+        if how_to_pickle not in ['pickle', 'joblib']:
```

Comment:
You already have this warning on line 250

Reply:
Yes, u are right , I should remove it